addon/components/layers/kml-layer.js

```diff
@@ -0,0 +1,152 @@
+            resolve(layer);
+          })
+          .on('error', (e) => {
+            reject(e.error || e);
+          });
+      });
+    }
+
+    if (kmlString) {
+      let layer = window.omnivore.kml.parse(kmlString);
+      this.set('_leafletObject', layer);
+      return layer;
+    }
+  },
+
+  /**
+    Handles 'flexberry-map:identify' event of leaflet map.
+
+    @method identify
+    @param {Object} e Event object.
+    @param {<a href="http://leafletjs.com/reference-1.0.0.html#latlngbounds">L.LatLngBounds</a>} options.boundingBox Bounds of identification area.
+    @param {<a href="http://leafletjs.com/reference-1.0.0.html#latlng">L.LatLng</a>} e.latlng Center of the bounding box.
+    @param {Object[]} layers Objects describing those layers which must be identified.
+    @param {Object[]} results Objects describing identification results.
+    Every result-object has the following structure: { layer: ..., features: [...] },
+    where 'layer' is metadata of layer related to identification result, features is array
+    containing (GeoJSON feature-objects)[http://geojson.org/geojson-spec.html#feature-objects]
+    or a promise returning such array.
+  */
+  identify(e) {
+    var bounds = new window.Terraformer.Primitive(e.polygonLayer.toGeoJSON());
+    let kmlLayer = this.get('_leafletObject');
+    return new Ember.RSVP.Promise((resolve, reject) => {
+      let features = Ember.A();
+      kmlLayer.eachLayer(function(layer) {
+        let feature = layer.toGeoJSON();
+        let primitive = new window.Terraformer.Primitive(feature.geometry);
+        let l = L.geoJSON(feature);
+
+        if (primitive instanceof window.Terraformer.Point ? primitive.within(bounds) : (primitive.intersects(bounds) || primitive.within(bounds))) {
+          feature.leafletLayer = l;
+          features.pushObject(feature);
+        }
+      });
+      resolve(features);
+    });
+  },
+
+  /**
+    Handles 'flexberry-map:search' event of leaflet map.
+
+    @method search
+    @param {Object} e Event object.
+    @param {<a href="http://leafletjs.com/reference-1.0.0.html#latlng">L.LatLng</a>} e.latlng Center of the search area.
+    @param {Object[]} layer Object describing layer that must be searched.
+    @param {Object} searchOptions Search options related to layer type.
+    @param {Object} results Hash containing search results.
+    @param {Object[]} results.features Array containing (GeoJSON feature-objects)[http://geojson.org/geojson-spec.html#feature-objects]
+    or a promise returning such array.
+  */
+  search(e) {
+    if (this.get('searchSettings.canBeSearched')) {
+      let that = this;
+      let kmlLayer = this.get('_leafletObject');
+
+      return new Ember.RSVP.Promise((resolve, reject) => {
+        let features = Ember.A();
+        kmlLayer.eachLayer(function(layer) {
+          if (features.length < e.searchOptions.maxResultsCount) {
+            let feature = layer.toGeoJSON();
+            let l = L.geoJSON(feature);
```

Comment:
cause be in if(contains) block, cause can be not use if feature dont correspond to search criteria.
feature variable may be so move to if(contains) block is layer variable contains properties.